pkg/controller/vitessshard/reconcile_rollout.go

```diff
@@ -0,0 +1,214 @@
+			planetscalev2.ClusterLabel:   vts.Labels[planetscalev2.ClusterLabel],
+			planetscalev2.KeyspaceLabel:  vts.Labels[planetscalev2.KeyspaceLabel],
+			planetscalev2.ShardLabel:     vts.Spec.KeyRange.SafeName(),
+		}).AsSelector(),
+	}
+	if err := r.client.List(ctx, listOpts, podList); err != nil {
+		return resultBuilder.Error(err)
+	}
+
+	scheduledTablets := false
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		if !rollout.Scheduled(pod) {
+			continue
+		}
+
+		scheduledTablets = true
+		if err := r.releaseTabletPod(ctx, log, pod); err != nil {
+			r.recorder.Eventf(vts, corev1.EventTypeWarning, "RollingRestartFailed", "tablet deletion failed")
+			return resultBuilder.Error(err)
+		}
+	}
+
+	// If we have no more scheduled tablets, unschedule the shard.
+	if !scheduledTablets {
+		if err := r.unscheduleObject(ctx, vts); err != nil {
+			return resultBuilder.Error(err)
+		}
+	}
+
+	return resultBuilder.Result()
+}
+
+func (r *ReconcileVitessShard) releaseTabletPod(ctx context.Context, log *logrus.Entry, pod *corev1.Pod) error {
+	// Release the pod to be recreated with updates.
+	if err := r.releaseObject(ctx,pod); err != nil {
+		return err
+	}
+
+	// TODO: Evict pods before deleting them.
+	if err := r.client.Delete(ctx, pod); err != nil {
+		return err
+	}
+	if err := r.waitObjectUpdated(ctx, pod); err != nil {
+		return err
+	}
+	if err := r.waitPodReady(ctx, pod); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *ReconcileVitessShard) waitObjectUpdated(ctx context.Context, obj runtime.Object) error {
+	// Wait for obj to have no scheduled updates.
+	for {
+		if err := regetObject(ctx, r.client, obj); err != nil {
+			// A NotFound error is ok. We wait for it to be recreated.
+			if apierrors.IsNotFound(err) {
+				time.Sleep(1 * time.Second)
+				continue
+			} else {
+				return err
+			}
+		}
+
+		if !rollout.Scheduled(objectMeta(obj)) {
+			// This is what we're waiting for.
+			return nil
+		}
+
+		// Keep waiting.
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (r *ReconcileVitessShard) waitPodReady(ctx context.Context, pod *corev1.Pod) error {
+	for {
+		if err := regetObject(ctx, r.client, pod); err != nil {
+			return err
+		}
+
+		if podIsReady(pod) {
+			// This is what we're waiting for.
+			return nil
+		}
+
+		// Keep waiting.
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (r *ReconcileVitessShard) releaseObject(ctx context.Context, obj runtime.Object) error {
+	return retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := regetObject(ctx, r.client, obj); err != nil {
+			return err
+		}
+		objMeta := objectMeta(obj)
+		if !rollout.Scheduled(objMeta) || rollout.Released(objMeta) {
+			// If there's nothing scheduled, or already released, we're done.
+			return nil
+		}
+		rollout.Release(objMeta)
+		return r.client.Update(ctx, obj)
+	})
+}
+
+func (r *ReconcileVitessShard) unscheduleObject(ctx context.Context, obj runtime.Object) error {
```

Comment:
Make this `unpropagateShard` and just pass `*planetscalev2.VitessShard` so we don't have to use the `objectMeta` casting. We only ever need this on that one type here.

Reply:
Done.